PRM-2567 Filter out new appeal statuses

AndrewRButcher opened the pull request.

LewisMitchell commented on app/services/FilterService.scala:
```diff
@@ -0,0 +1,139 @@
+
+  def filterEstimatedLPP1DuringPeriodOfFamiliarisation(penaltiesDetails: GetPenaltyDetails, callingClass: String, function: String, vrn: String): GetPenaltyDetails = {
+    if (penaltiesDetails.latePaymentPenalty.nonEmpty) {
+      val filtered: Option[Seq[LPPDetails]] = filterEstimatedLPP1(penaltiesDetails)
+      val numberOfFiltered = numberOfFilteredLPPs(filtered, penaltiesDetails)
```
Setting these to filteredLPPs would help with readability and maybe saying that they are filtered due to being LPP1's in POF

LewisMitchell commented on app/services/FilterService.scala:
```diff
@@ -0,0 +1,139 @@
+  }
+
+  def filterPenaltiesWith9xAppealStatus(penaltiesDetails: GetPenaltyDetails)(implicit callingClass: String, function: String, vrn: String): GetPenaltyDetails = {
+    val filteredLSPs: Option[Seq[LSPDetails]] = if (penaltiesDetails.lateSubmissionPenalty.nonEmpty) filterLSPWith9xAppealStatus(penaltiesDetails) else None
```
Why use if? Could we not just call the filterLSPWith9xAppealStatus func directly?

LewisMitchell commented on app/services/FilterService.scala:
```diff
@@ -0,0 +1,139 @@
+  def filterPenaltiesWith9xAppealStatus(penaltiesDetails: GetPenaltyDetails)(implicit callingClass: String, function: String, vrn: String): GetPenaltyDetails = {
+    val filteredLSPs: Option[Seq[LSPDetails]] = if (penaltiesDetails.lateSubmissionPenalty.nonEmpty) filterLSPWith9xAppealStatus(penaltiesDetails) else None
+    val noOfFilteredLSPs:Int = numberOfFilteredLSPs(filteredLSPs, penaltiesDetails)
+    val filteredLPPs: Option[Seq[LPPDetails]] = if (penaltiesDetails.latePaymentPenalty.nonEmpty) filterLPPWith9xAppealStatus(penaltiesDetails) else None
```
Why use if? Could we not just call the filterLPPWith9xAppealStatus func directly?

LewisMitchell commented on app/services/FilterService.scala:
```diff
@@ -0,0 +1,139 @@
+    val noOfFilteredLPPs: Int = numberOfFilteredLPPs(filteredLPPs, penaltiesDetails)
+
+    (noOfFilteredLSPs, noOfFilteredLPPs) match {
+      case (x , 0) if x > 0 => penaltiesDetails.copy(lateSubmissionPenalty = prepareLateSubmissionPenaltiesAfterFilter(penaltiesDetails, filteredLSPs, noOfFilteredLSPs, vrn, callingClass, function))
+      case (0, x) if x > 0 =>
+        penaltiesDetails.copy(latePaymentPenalty = prepareLatePaymentPenaltiesAfterFilter(filteredLPPs, noOfFilteredLPPs, vrn, callingClass, function))
+      case (x, y) if x > 0 && y > 0 =>
+        val newLasteSubmissions: Option[LateSubmissionPenalty] = prepareLateSubmissionPenaltiesAfterFilter(penaltiesDetails, filteredLSPs, noOfFilteredLSPs, vrn, callingClass, function)
+        val newLastePaymentPenalties: Option[LatePaymentPenalty] = prepareLatePaymentPenaltiesAfterFilter(filteredLPPs, noOfFilteredLPPs, vrn, callingClass, function)
+        penaltiesDetails.copy(lateSubmissionPenalty = newLasteSubmissions, latePaymentPenalty = newLastePaymentPenalties)
+      case _ =>
```
This code is quite convoluted, is there not a generic way we can return these LSPs and LPPs. I'm conscious of the effect this will have on performance

LewisMitchell commented on app/services/FilterService.scala:
```diff
@@ -0,0 +1,139 @@
+
+  private def numberOfFilteredLPPs(filteredLPPs: Option[Seq[LPPDetails]], penaltiesDetails: GetPenaltyDetails): Int = {
+    if (penaltiesDetails.latePaymentPenalty.nonEmpty) {
+      penaltiesDetails.latePaymentPenalty.get.details.get.size - filteredLPPs.get.size
```
Maybe worth having a check on filteredLPPs being non empty

LewisMitchell commented on app/services/FilterService.scala:
```diff
@@ -0,0 +1,139 @@
+    }
+  }
+
+  private def filterLPPWith9xAppealStatus(penaltiesDetails: GetPenaltyDetails): Option[Seq[LPPDetails]] = {
+    penaltiesDetails.latePaymentPenalty.flatMap(
+      _.details.map(latePaymentPenalties => latePaymentPenalties.filterNot(lpp => {
+        lpp.appealInformation.nonEmpty && lpp.appealInformation.get.exists(appealInfo => appealInfo.appealStatus.nonEmpty && (appealInfo.appealStatus.get.equals(AppealStatusEnum.AppealRejectedChargeAlreadyReversed) || appealInfo.appealStatus.get.equals(AppealStatusEnum.AppealUpheldChargeAlreadyReversed) || appealInfo.appealStatus.get.equals(AppealStatusEnum.AppealRejectedPointAlreadyRemoved) || a ...
+      })
+      )
+    )
+  }
+
+  private def filterLSPWith9xAppealStatus(penaltiesDetails: GetPenaltyDetails): Option[Seq[LSPDetails]] = {
+    penaltiesDetails.lateSubmissionPenalty.map(_.details.filterNot(lsp =>
+      lsp.appealInformation.nonEmpty && lsp.appealInformation.get.exists(appealInfo => appealInfo.appealStatus.nonEmpty && (appealInfo.appealStatus.get.equals(AppealStatusEnum.AppealRejectedChargeAlreadyReversed) || appealInfo.appealStatus.get.equals(AppealStatusEnum.AppealUpheldChargeAlreadyReversed) || appealInfo.appealStatus.get.equals(AppealStatusEnum.AppealRejectedPointAlreadyRemoved) || app ...
+  }
```
Is there not a way we could combine these methods?

LewisMitchell commented on test/base/LSPDetailsBase.scala:
```diff
@@ -0,0 +1,57 @@
+    chargeAmount = None
+  )
+
+  def lspPointDetailsWithAppealStatus(status: String):LSPDetails = lspPointDetails.copy(appealInformation = Some(Seq(AppealInformationType(appealStatus = Some(AppealStatusEnum.withName(status)), appealLevel = Some(AppealLevelEnum.HMRC)))))
```
If this were an appeal point, then the penaltyStatus would also change

LewisMitchell commented on test/base/LSPDetailsBase.scala:
```diff
@@ -0,0 +1,57 @@
+
+  val lspThresholdDetails: LSPDetails = lspPointDetails.copy(penaltyCategory = LSPPenaltyCategoryEnum.Threshold, chargeAmount = Some(200), chargeOutstandingAmount = Some(200), chargeDueDate = Some(LocalDate.now()))
+
+  def lspThresholdDetailsWithAppealStatus(status: String):LSPDetails = lspThresholdDetails.copy(appealInformation = Some(Seq(AppealInformationType(appealStatus = Some(AppealStatusEnum.withName(status)), appealLevel = Some(AppealLevelEnum.HMRC)))))
```
If this were an appeal point, then the penaltyStatus would also change

LewisMitchell replied:
expiryReason would be APP

LewisMitchell commented on test/base/LSPDetailsBase.scala:
```diff
@@ -0,0 +1,57 @@
+
+  val lspChargeDetails: LSPDetails = lspPointDetails.copy(penaltyCategory = LSPPenaltyCategoryEnum.Charge, chargeAmount = Some(200), chargeOutstandingAmount = Some(200), chargeDueDate = Some(LocalDate.now()))
+
+  def lspChargeDetailsWithAppealStatus(status: String): LSPDetails = lspChargeDetails.copy(appealInformation = Some(Seq(AppealInformationType(appealStatus = Some(AppealStatusEnum.withName(status)), appealLevel = Some(AppealLevelEnum.HMRC)))))
```
If this were an appeal point, then the penaltyStatus would also change

LewisMitchell commented on test/services/FilterServiceSpec.scala:
```diff
@@ -0,0 +1,282 @@
+  }
+
+  "filterEstimatedLPP1DuringPeriodOfFamiliarisation" should {
+    "filter LPP1s with a principle charge due within the filtering window" in {
```
```suggestion
    "filter LPP1s with a principal charge due within the filtering window" in {
```

LewisMitchell commented on test/services/FilterServiceSpec.scala:
```diff
@@ -0,0 +1,282 @@
+      result shouldBe expectedResult
+    }
+
+    "remove all LPP1s with a principle charge due within the filtering window" in {
```
```suggestion
    "remove all LPP1s with a principal charge due within the filtering window" in {
```

LewisMitchell commented on test/services/FilterServiceSpec.scala:
```diff
@@ -0,0 +1,282 @@
+      result shouldBe expectedResult
+    }
+
+    "remove no LPP1s when their principle charges are NOT due in the filtering window" in {
```
```suggestion
    "remove no LPP1s when their principal charges are NOT due in the filtering window" in {
```

LewisMitchell replied:
To be fixed during refactor

LewisMitchell commented on test/services/FilterServiceSpec.scala:
```diff
@@ -0,0 +1,282 @@
+  }
+
+  "filterPenaltiesWith9xAppealStatus" should {
+    "filter LSPs with am appeal state of 91,92, 93 or 94" in {
```
```suggestion
    "filter LSPs with an appeal status of 91,92, 93 or 94" in {
```

LewisMitchell commented on test/services/FilterServiceSpec.scala:
```diff
@@ -0,0 +1,282 @@
+
+  "filterPenaltiesWith9xAppealStatus" should {
+    "filter LSPs with am appeal state of 91,92, 93 or 94" in {
+      val LspSummary = LSPSummary(
```
```suggestion
      val lspSummary = LSPSummary(
```

LewisMitchell commented on test/services/FilterServiceSpec.scala:
```diff
@@ -0,0 +1,282 @@
+        expectedResult.lateSubmissionPenalty.get.details
+    }
+
+    "filter LPPs with am appeal state of 91,92, 93 or 94" in {
```
```suggestion
    "filter LPPs with an appeal status of 91, 92, 93 or 94" in {
```

LewisMitchell commented on test/services/FilterServiceSpec.scala:
```diff
@@ -0,0 +1,282 @@
+  }
+
+
+  "filter LSPs and LPPs with am appeal state of 91,92, 93 or 94" in {
```
```suggestion
  "filter LSPs and LPPs with an appeal status of 91, 92, 93 or 94" in {
```

LewisMitchell commented on test/services/FilterServiceSpec.scala:
```diff
@@ -0,0 +1,282 @@
+
+
+  "filter LSPs and LPPs with am appeal state of 91,92, 93 or 94" in {
+    val LspSummary = LSPSummary(
```
```suggestion
    val lspSummary = LSPSummary(
```

LewisMitchell commented on test/services/FilterServiceSpec.scala:
```diff
@@ -0,0 +1,282 @@
+  }
+
+  "Not filter LSPs and LPPs" in {
+    val LspSummary = LSPSummary(
```
```suggestion
    val lspSummary = LSPSummary(
```

AndrewRButcher commented:
Address PR moments ✅

AndrewRButcher commented:
Added feature switch + addressed PR comments
PRs for FS in QA and STAGING
https://github.com/hmrc/app-config-qa/pull/18596
https://github.com/hmrc/app-config-staging/pull/12840

LewisMitchell commented on app/services/FilterService.scala:
```diff
@@ -0,0 +1,136 @@
+  def filterEstimatedLPP1DuringPeriodOfFamiliarisation(penaltiesDetails: GetPenaltyDetails, callingClass: String, function: String, vrn: String): GetPenaltyDetails = {
+    if (penaltiesDetails.latePaymentPenalty.nonEmpty) {
+      val filteredLPPs: Option[Seq[LPPDetails]] = filterEstimatedLPP1DuringPOF(penaltiesDetails)
+      val numberOfFiltered = numberOfFilteredLPPs(filteredLPPs, penaltiesDetails)
```
Still valid, to be fixed during refactor

LewisMitchell commented on test/base/LSPDetailsBase.scala:
```diff
@@ -0,0 +1,57 @@
+    chargeAmount = None
+  )
+
+  def lspPointDetailsWithAppealStatus(status: String):LSPDetails = lspPointDetails.copy(appealInformation = Some(Seq(AppealInformationType(appealStatus = Some(AppealStatusEnum.withName(status)), appealLevel = Some(AppealLevelEnum.HMRC)))), expiryReason = Some(ExpiryReasonEnum.Appeal), penaltyStatus = LSPPenaltyStatusEnum.Inactive)
```
The penalty should only be inactive when the appeal status is B

LewisMitchell commented on test/base/LSPDetailsBase.scala:
```diff
@@ -0,0 +1,57 @@
+
+  val lspChargeDetails: LSPDetails = lspPointDetails.copy(penaltyCategory = LSPPenaltyCategoryEnum.Charge, chargeAmount = Some(200), chargeOutstandingAmount = Some(200), chargeDueDate = Some(LocalDate.now()))
+
+  def lspChargeDetailsWithAppealStatus(status: String): LSPDetails = lspChargeDetails.copy(appealInformation = Some(Seq(AppealInformationType(appealStatus = Some(AppealStatusEnum.withName(status)), appealLevel = Some(AppealLevelEnum.HMRC)))), expiryReason = Some(ExpiryReasonEnum.Appeal), penaltyStatus = LSPPenaltyStatusEnum.Inactive)
```
The penalty should only be inactive when the appeal status is B

LewisMitchell commented on test/services/FilterServiceSpec.scala:
```diff
@@ -0,0 +1,312 @@
+  sys.props -= featureSwitching.ESTIMATED_LPP1_FILTER_END_DATE
+
+  val mockAppConfig: AppConfig = new AppConfig(mockConfig, mockServicesConfig)
+  val filter = new FilterService
```
To be fixed during refactor

AndrewRButcher commented:
AT PR required https://github.com/hmrc/penalties-acceptance-tests/pull/265
